### PR TITLE
[CodeCompletion] Suggest synthesized declarations from macros

### PR DESCRIPTION
**Pull request by rintaro**
* Don't invalidate the lookup cache in `getOrCreateSynthesizedFile()` Adding a synthesized file itself doesn't introduce any decls. Instead, we should invalidate it right after the actual declrations are added in the file
* Remove `SourceLookupCache::invalidate()` method. It was just used right before the destruction. It was just not necessary
* Include auxiliary decls in `SourceLookupCache::lookupVisibleDecls()` Previously, global symbol completion didn't include decls synthesized by peer macros or freestanding decl macros
* Include "auxiliary decls" in visible member lookup, and visible local decl lookup
* Hide macro unique names

rdar://110535113

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -3827,6 +3827,10 @@ bool ValueDecl::shouldHideFromEditor() const {
       getBaseIdentifier().str().startswith("$__"))
     return true;
 
+  // Macro unique names are only intended to be used inside the expanded code.
+  if (MacroDecl::isUniqueMacroName(getBaseName()))
```
Comment (bnbarham) [inline lib/AST/Decl.cpp:3831]: Nice 👍

File: lib/Sema/LookupVisibleDecls.cpp [hunk 2]
```diff
@@ -634,16 +641,6 @@ static void synthesizeMemberDeclsForLookup(NominalTypeDecl *NTD,
                             false);
   }
 
-  // Expand peer macros.
-  for (auto *member : NTD->getMembers()) {
-    if (!ctx.evaluator.hasActiveRequest(ExpandPeerMacroRequest{member})) {
-      (void)evaluateOrDefault(
-          ctx.evaluator,
-          ExpandPeerMacroRequest{member},
-          {});
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply lib/Sema/LookupVisibleDecls.cpp:643]: What calls `visitAuxiliaryDecls` for members in regular type checking?

File: test/IDE/complete_macros_expanded.swift [hunk 3]
```diff
@@ -0,0 +1,79 @@
+// REQUIRES: swift_swift_parser
+
+// RUN: %empty-directory(%t)
+// RUN: mkdir -p %t/plugins
+
+//##-- Prepare the macro plugin.
+// RUN: %host-build-swift -swift-version 5 -emit-library -o %t/plugins/%target-library-name(MacroDefinition) -module-name=MacroDefinition %S/../Macros/Inputs/syntax_macro_definitions.swift -g -no-toolchain-stdlib-rpath
+
+// RUN: %target-swift-ide-test -batch-code-completion -source-filename %s -filecheck %raw-FileCheck -completion-output-dir %t -plugin-path %t/plugins -parse-as-library 
+
```
Comment (bnbarham) [inline test/IDE/complete_macros_expanded.swift:10]: Since `arbitrary` goes down fairly different paths to named, possibly worth adding tests for it as well?

File: lib/AST/NameLookup.cpp [hunk 4]
```diff
@@ -3875,20 +3876,45 @@ void FindLocalVal::visitBraceStmt(BraceStmt *S, bool isTopLevelCode) {
...
     // If we have a SingleValueStmtExpr, there may be local bindings in the
     // wrapped statement.
     if (auto *E = elem.dyn_cast<Expr *>()) {
       if (auto *SVE = dyn_cast<SingleValueStmtExpr>(E))
         visit(SVE->getStmt());
+      continue;
     }
+
     if (auto *S = elem.dyn_cast<Stmt*>())
       visit(S);
```
Comment (ahoppen) [inline lib/AST/NameLookup.cpp:3891]: Very nit-picky: Wouldn’t this be cleaner represented as

```suggestion
    } else if (auto *S = elem.dyn_cast<Stmt*>()) {
      visit(S);
    }
```

File: lib/AST/NameLookup.cpp [hunk 5]
```diff
@@ -3875,20 +3876,45 @@ void FindLocalVal::visitBraceStmt(BraceStmt *S, bool isTopLevelCode) {
...
         checkValueDecl(VD, DeclVisibilityKind::LocalVariable);
+      // FIXME: Recursively call `visitDecl` to handle nested macros.
+    });
+  };
+
+  for (auto elem : S->getElements()) {
+    if (auto *E = elem.dyn_cast<Expr *>()) {
+      // 'MacroExpansionExpr' at code-item position may introduce value decls.
+      // NOTE: the expression must be type checked.
+      // FIXME: In code-completion local expressions are _not_ type checked.
```
Comment (ahoppen) [inline lib/AST/NameLookup.cpp:3908]: How do we get completions from macro expansion declarations if the corresponding `MacroExpansionExpr` hasn’t been type-checked?

File: lib/SILOptimizer/Differentiation/LinearMapInfo.cpp [hunk 6]
```diff
@@ -117,6 +117,7 @@ LinearMapInfo::createBranchingTraceDecl(SILBasicBlock *originalBB,
     branchingTraceDecl->setAccess(AccessLevel::Internal);
   }
   file.addTopLevelDecl(branchingTraceDecl);
+  file.getParentModule()->clearLookupCache();
```
Comment (ahoppen) [inline lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:120]: Out of curiosity: Why is this needed?

File: test/IDE/complete_macros_expanded.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+// MEMBER_STATIC-NOT: _peer
+}
+
+func testLocal() {
+  #defineDeclsWithKnownNames
+
+  @PeerWithSuffix func localFunc() {}
+
+  do {
+    #^LOCAL?skip=FIXME^#
```
Comment (ahoppen) [inline test/IDE/complete_macros_expanded.swift:69]: Change this to an `xfail` so we don’t forget to change it if it’s fixed?
```suggestion
    #^LOCAL?xfail=FIXME^#
```

**Review by ahoppen — APPROVED** [review]
Thank you 🙏🏽

File: lib/SILOptimizer/Differentiation/LinearMapInfo.cpp (hunk 6 above)
Comment (rintaro) [reply lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:120]: We need to clear cache after something is added to the top-level, otherwise `ModuleDecl::lookupValue()` won't pick up the added decl.
Previously, this was called in `getOrCreateSynthesizedFile()`, but it was technically a wrong place because the user calls `ModuleDecl::lookupValue()` before adding a top-level decl in it, it'd break.

File: lib/Sema/LookupVisibleDecls.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rintaro) [reply lib/Sema/LookupVisibleDecls.cpp:643]: I _think_ it's `populateLookupTableEntryFromMacroExpansions` called from `DirectLookupRequest::evaluate()`

File: test/IDE/complete_macros_expanded.swift (hunk 3 above)
Comment (rintaro) [reply test/IDE/complete_macros_expanded.swift:10]: Apparently `arbitrary` at top-level is not really supported:
If we add `names: arbitrary`
* `error: 'declaration' macros are not allowed to introduce arbitrary names at global scope`
* `error: 'peer' macros are not allowed to introduce arbitrary names at global scope`
And the macro decl becomes `isInvalid()`, and the expansion is not performed.

If we omit `names` from the macro decl:
* `error: declaration name '<name>' is not covered by macro '<macro name>'`
Introduced values are found and type checked, but since it's `error`, it doesn't built.

I feel like we should remove 
```cpp
   unexpandedDecls.append(TopLevelArbitraryMacros.begin(),
                          TopLevelArbitraryMacros.end());
```
From `lookupVisibleDecls()`

Comment (bnbarham) [reply test/IDE/complete_macros_expanded.swift:10]: I was thinking within a struct

Comment (rintaro) [reply test/IDE/complete_macros_expanded.swift:10]: For non top level cases, there is not much different code paths between arbitrary and named, but sure.

File: lib/AST/NameLookup.cpp (hunk 4 above)
Comment (rintaro) [reply lib/AST/NameLookup.cpp]: I actually don't like `else if` branches for iterating union cases, unless the branch contents are almost the same or one line code. For example, this is OK,
```cpp
void ASTNode::walk(ASTWalker &Walker) {
  if (auto *E = this->dyn_cast<Expr*>())
    E->walk(Walker);
  else if (auto *S = this->dyn_cast<Stmt*>())
    S->walk(Walker);
  else if (auto *D = this->dyn_cast<Decl*>())
    D->walk(Walker);
  else if ...
```
But for this specific case I prefer `if ... continue`.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/AST/NameLookup.cpp (hunk 4 above)
Comment (ahoppen) [reply lib/AST/NameLookup.cpp]: OK, if you’ve got a preference, that’s fine with me.

File: lib/SILOptimizer/Differentiation/LinearMapInfo.cpp (hunk 6 above)
Comment (ahoppen) [reply lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:120]: Thanks for the explanation. If you want, you can add a comment with it for future reference.

File: test/IDE/complete_macros_expanded.swift (hunk 7 above)
Comment (ahoppen) [reply test/IDE/complete_macros_expanded.swift:86]: Ah, OK. Makes sense.
